### PR TITLE
[WIP] Squad refactor

### PR DESCRIPTION
**Pull request by LysandreJik**
This PR aims to refactor SQuAD to make it usable with all models with question answering heads, and without having to build the entire tokenization pipeline as it is currently done.

- It is based on processors that manage data, similarly to the GLUE processors. The two new processors are `SquadV1Processor` and `SquadV2Processor`. They'll probably be merged into a single `SquadProcessor` as the difference between the two versions is minimal.
- It leverages powerful abstractions made for the `run_glue` refactor a few months ago that greatly simplified the tokenization pipeline
- It can be interfaced with the package `tensorflow_datasets`.
- It better respects the library-wide naming, with `attention_mask` instead of `input_mask` and `token_type_ids` instead of `segment_ids`, among others.
- Introduces padding to `encode` and `encode_plus`, alongside tests.

It is still a work on progress but some aspects of it are working.

### Left to do

- [x] Add the processors to `__init__.py`
- [x] Patch the evaluation so that it leverages the current interface
- [x] Patch the evaluation so that it may work with tfds
- [x] Modify the run arguments to reflect the changes
- [x] Remove the `only_first` argument which would only be used for testing
- [x] Update tests running the `run_squad.py` script
- [x] Include the padding location in the tokenizers and reflect the changes in the feature converter
- [x] Test that all current models can train and evaluate (BERT, RoBERTa, XLNet, XLM)
- [x] Add the last models (DistilBERT, ALBERT, ...)
- [x] Return datasets (maybe only pytorch TensorDataset for now)
- [x] Documentation
- [x] Short examples showcasing the simply usage in the processors section.
- [x] Patch the evaluation for impossible questions

### Running sample

Here's the major difference from the user's perspective. Initially, to obtain the examples which were then converted to features, the user had to do as follows (taken from the current `run_squad.py`), which only works for BERT/XLNet/DistilBERT/ALBERT:

```py

examples = read_squad_examples(
    input_file=input_file,
    is_training=not evaluate,
    version_2_with_negative=args.version_2_with_negative
)
     
features = convert_examples_to_features(
    examples=examples,
    tokenizer=tokenizer,
    max_seq_length=args.max_seq_length,
    doc_stride=args.doc_stride,
    max_query_length=args.max_query_length,
    is_training=not evaluate,
    cls_token_segment_id=2 if args.model_type in ['xlnet'] else 0,
    pad_token_segment_id=3 if args.model_type in ['xlnet'] else 0,
    cls_token_at_end=True if args.model_type in ['xlnet'] else False,
    sequence_a_is_doc=True if args.model_type in ['xlnet'] else False
)
```
In order to obtain the exact same results, the user now has to do as follows, which will be completely model independant once the `sequence_a_is_doc` is integrated in our sequence pair tokenization methods:

```py
processor = SquadV1Processor()
examples = processor.get_dev_examples("examples/squad") if evaluate else processor.get_train_examples("examples/squad")
features = squad_convert_examples_to_features(
    examples=examples,
    tokenizer=tokenizer,
    max_seq_length=args.max_seq_length,
    doc_stride=args.doc_stride,
    max_query_length=args.max_query_length,
    is_training=not evaluate,
    sequence_a_is_doc=True if args.model_type in ['xlnet'] else False
)
```

The same can be done by using TFDS instead, removing the need to specify a file. The two initial lines now become:

```py
tfds_examples = tensorflow_datasets.load("squad")["validation"] if evaluate else tensorflow_datasets.load("squad")["train"] 
examples = SquadV1Processor().get_examples_from_dataset(tfds_examples)
```

### CODE REVIEW

File: transformers/tokenization_utils.py [hunk 1]
```diff
@@ -718,6 +729,12 @@ def encode(self,
                 - 'only_first': Only truncate the first sequence
                 - 'only_second': Only truncate the second sequence
                 - 'do_not_truncate': Does not truncate (raise an error if the input sequence is longer than max_length)
+            padding_strategy: if set to a strategy, the returned sequences will be padded according to the model's 
+                padding index, up to their max length. If no max length is specified, no padding is done.
```
Comment (thomwolf) [inline transformers/tokenization_utils.py]: In the most general case for the user, we may want to allow passing a batch of sequence and just have them padded to the maximal sequence length in the batch (or probably the min between the model's max length and the maximal sequence length in the batch).
Should we prepare for that?

Comment (thomwolf) [reply transformers/tokenization_utils.py]: I feel like we should do padding up to `self.max_length` if padding is asked even when no explicit `max_length` is provided to the method.

File: transformers/tokenization_utils.py [hunk 2]
```diff
@@ -190,6 +190,11 @@ def pad_token_id(self):
         """ Id of the padding token in the vocabulary. Log an error if used while not having been set. """
         return self.convert_tokens_to_ids(self.pad_token)
 
+    @property
+    def pad_token_type_id(self):
+        """ Id of the padding token in the vocabulary. Log an error if used while not having been set. """
```
Comment (thomwolf) [inline transformers/tokenization_utils.py]: we can remove `Log an error if used while not having been set` I think

**Review by thomwolf — COMMENTED** [review]
That's really a lot cleaner, I like it very much.
Great work @LysandreJik!

File: transformers/tokenization_utils.py (hunk 1 above)
Comment (LysandreJik) [reply transformers/tokenization_utils.py]: Regarding the first comment, we don't currently manage sequence batching so it would require implementing an `encode_batch` method or similar beforehand, wouldn't it? As of right now, it shouldn't be too hard for the user to identify the longer sequence in its batches, but it would be a good idea to ease the process of tokenizing batches.

Totally agree with your second comment, this makes sense!

File: transformers/tokenization_utils.py (hunk 2 above)
Comment (LysandreJik) [reply transformers/tokenization_utils.py]: Very true!

File: docs/source/main_classes/processors.rst [hunk 3]
```diff
@@ -55,4 +55,81 @@ Example usage
 ^^^^^^^^^^^^^^^^^^^^^^^^^
 
 An example using these processors is given in the
-`run_glue.py <https://github.com/huggingface/pytorch-transformers/blob/master/examples/run_glue.py>`__ script.
\ No newline at end of file
+`run_glue.py <https://github.com/huggingface/pytorch-transformers/blob/master/examples/run_glue.py>`__ script.
```
Comment (stefan-it) [inline docs/source/main_classes/processors.rst]:
```suggestion
`run_glue.py <https://github.com/huggingface/transformers/blob/master/examples/run_glue.py>`__ script.
```

File: docs/source/main_classes/processors.rst [hunk 4]
```diff
@@ -55,4 +55,81 @@ Example usage
...
+        tokenizer=tokenizer,
+        max_seq_length=max_seq_length,
+        doc_stride=args.doc_stride,
+        max_query_length=max_query_length,
+        is_training=not evaluate,
+    )
+
+
+Another example using these processors is given in the
+`run_squad.py <https://github.com/huggingface/pytorch-transformers/blob/master/examples/run_squad.py>`__ script.
```
Comment (stefan-it) [inline docs/source/main_classes/processors.rst]:
```suggestion
`run_squad.py <https://github.com/huggingface/transformers/blob/master/examples/run_squad.py>`__ script.
```

**Comment by LysandreJik** [comment]
Should be about ready to merge now. I'm reproducing the results paper results on XLNet and BERT to make sure it works as expected, then I'll rebase and call for review.

File: transformers/tokenization_utils.py [hunk 5]
```diff
@@ -214,10 +221,14 @@ def __init__(self, max_len=None, **kwargs):
         self._pad_token = None
         self._cls_token = None
         self._mask_token = None
+        self._pad_token_type_id = 0
         self._additional_special_tokens = []
 
         self.max_len = max_len if max_len is not None else int(1e12)
 
+        # Padding side is right by default and over-riden in subclsses. If specified in the kwargs, it is changed.
```
Comment (thomwolf) [inline transformers/tokenization_utils.py]: 👍 nice

Comment (thomwolf) [reply transformers/tokenization_utils.py]: Let's add it to the class docstring later to have it in the doc as well.

File: transformers/tokenization_utils.py [hunk 6]
```diff
@@ -951,7 +995,43 @@ def prepare_for_model(self, ids, pair_ids=None, max_length=None, add_special_tok
...
                            "indexing errors".format(len(ids), self.max_len))
+                           
+        needs_to_be_padded = pad_to_max_length and (
+            max_length and len(encoded_inputs["input_ids"]) < max_length
+            or 
+            max_length is None and len(encoded_inputs["input_ids"]) < self.max_len and self.max_len <= 10000
+        )
+
+        if pad_to_max_length and max_length is None and self.max_len > 10000:
+            logger.warning("Sequence can't be padded as the maximum  ")
```
Comment (thomwolf) [inline transformers/tokenization_utils.py]: Warning seems incomplete

**Comment by thomwolf** [comment]
Ok this is great, merging!

@LysandreJik do you want to have a quick look at the two comments I wrote about doc/docstring and the incomplete warning?

Merging now because these are small tweak and @mfuntowicz need this PR for his PR so I'll let you push a doc commit on master directly maybe.

File: transformers/tokenization_utils.py [hunk 7]
```diff
@@ -741,6 +757,12 @@ def encode(self,
                 - 'only_first': Only truncate the first sequence
                 - 'only_second': Only truncate the second sequence
                 - 'do_not_truncate': Does not truncate (raise an error if the input sequence is longer than max_length)
+            pad_to_max_length: if set to True, the returned sequences will be padded according to the model's padding side and
+                padding index, up to their max length. If no max length is specified, the padding is done up to the model's max length.
+                The tokenizer padding sides are handled by the following strings:
```
Comment (LysandreJik) [inline transformers/tokenization_utils.py]: @thomwolf Is that what you meant by class substring?
